README.md
# Cordova Android Utility Plugin [![npm version](https://badge.fury.io/js/cordova-plugin-android-utility.svg)](https://badge.fury.io/js/cordova-plugin-android-utility)

This plugin is built as a general tool utility to perform relatively small tasks on the native android side.

---

## Contents

- [Setup](#Setup)
- [Methods](#methods)
  - [getVersionInfo()](#getVersionInfo)
  - [getNotificationChannels()](#getNotificationChannels)
  - [getNotificationChannel()](#getNotificationChannel)
  - [getNotificationSettings()](#getNotificationSettings)
  - [openNotificationSettings()](#openNotificationSettings)
  - [getSimInfo()](#getSimInfo)
- [Questions](#Questions?)
- [License](#License)

## Setup

```shell
cordova plugin add cordova-plugin-android-utility

# or

cordova plugin add https://github.com/DavidBriglio/cordova-plugin-android-utility
```

---

## Methods

### getVersionInfo

`Usage: cordova.plugins.android.utility.getVersionInfo()`

This method will return information on the android version.

Returns Promise with:

| Index | Type |
|-------|------|
| sdk | Integer |
| baseOs | String |
| codeName | String |
| incremental | String |
| previewSdk | Integer |
| release | String |
| securityPatch | String |

Example Usage:

```javascript
cordova.plugins.android.utility.getVersionInfo()
  .then(info => {
    // Show the android official release number and the sdk number
    console.log(`Android ${info.release} (SDK ${info.sdk})`)
  })
  .catch(message => console.log('Error: ' + message))
```

### getNotificationChannels

*Android 26+*

`Usage: cordova.plugins.android.utility.getNotificationChannels()`

This method will return a promise with an array of notification channel id strings.

Example Usage:

```javascript
cordova.plugins.android.utility.getNotificationChannels()
  .then(channels => {
    // List all channels
    for (let i in channels) {
      console.log(`Channel ${i}: ${channels[i]}`)
    }
  })
  .catch(message => console.log('Error: ' + message))
```

### getNotificationChannel

*Android 26+*

`Usage: cordova.plugins.android.utility.getNotificationChannel([string: channel id])`

This method will return a promise with all information about the specified channel id.

[Android Documentation](https://developer.android.com/reference/android/app/NotificationChannel)

Returns Promise with:
| Index | Type | Description |
|-------|------|-------------|
| canBypassDnd | Boolean | True if the notification bypasses do not disturb. |
| canShowBadge | Boolean | True if the notification will show a badge when active. |
| description | String | Description of the channel. |
| group | String | Channel group. |
| id | String | Channel ID. |
| importance | Integer | Integer representing channel importance. Refer to android documentation for value representations. |
| lightColour | Integer | Integer representation of the light colour shown when `showsLights` is true. |
| lockscreenVisibility | Integer | Integer value representing visibility of the notification on the lock screen. Refer to the android documentation for value representations. |
| name | String | Name of the channel. |
| sound | String | URI of the sound being used. |
| vibrationPattern | Array | Vibration pattern array. |
| showsLights | Boolean | Can show LED light when notification is active. |
| vibration | Boolean | True if notification will cause a vibration. |
| summary | String | The `toString` method result of the channel. |

Example Usage:

```javascript
cordova.plugins.android.utility.getNotificationChannel('TEST_CHANNEL_1')
  .then(channel => {
    // Show some of the channel information
    console.log('This channel is: ' + channel.name)
    console.log('This channel has vibration: ' + (channel.vibration ? 'Enabled' : 'Disabled'))
    console.log('This channel uses the sound: ' + channel.sound)
  })
  .catch(message => console.log('Error: ' + message))
```

### getNotificationSettings

`Usage: cordova.plugins.android.utility.getNotificationSettings()`

Obtain the base notification settings for the application. Use this for global settings and with API 25-.

Returns Promise with:

| Index | Type | Description |
|-------|------|-------------|
| enabled | Boolean | True if notifications are enabled for the app. |
| importance | Integer | Integer value representing importance of notifications in the app. (API 25-) |

```javascript
cordova.plugins.android.utility.getNotificationSettings()
  .then(settings => {
    // Show the global notification settings for the app
    console.log('Notifications are: ' + (settings.enabled ? 'Enabled' : 'Disabled'))
    console.log('Importance Level: ' + settings.importance)
  })
  .catch(message => console.log('Error: ' + message))
```

### openNotificationSettings

`Usage: cordova.plugins.android.utility.openNotificationSettings()`

Open the Android OS notification settings page corresponding to the current application. This method takes no parameters and does not return any values.

### getSimInfo

`Usage: cordova.plugins.android.utility.getSimInfo()`

Obtain the sim card information including the device phone number, imei and sim serial number.

| Index | Permission Required |
|-------|---------------------|
| phoneNumber | android.permission.READ_PHONE_STATE |
| imei | android.permission.READ_PRIVILEGED_PHONE_STATE |
| serial | android.permission.READ_PRIVILEGED_PHONE_STATE |

```javascript
// NOTE: Proper permissions must be requested first
cordova.plugins.android.utility.getSimInfo()
  .then(info => console.log(`Phone Number: ${info.phoneNumber}, IMEI: ${info.imei}, Serial#: ${info.serial}`))
  .catch(message => console.log(`Error Message: ${message}`))
```

---

## Questions?

Please feel free to open an issue or make a pull request!

---

## License

MIT - Please see the LICENSE file for details.

---

David Briglio (2020)


CHANGELOG.md
# Changelog

## 1.3.0 - 30-03-2021

- Added `getSimInfo` method

## 1.2.0 - 28-07-2020

- Added `openNotificationSettings` method

## 1.1.0 - 14-07-2020

- Added `getVersionInfo` method
- Added `getNotificationSettings` method

## 1.0.0 - 14-07-2020

- Initial release


www/utility.js
var exec = require('cordova/exec');

module.exports = {
  getNotificationChannel: id => new Promise((resolve, reject) => exec(resolve, reject, 'UtilityPlugin', 'getNotificationChannel', [id || ''])),
  getNotificationChannels: () => new Promise((resolve, reject) => exec(resolve, reject, 'UtilityPlugin', 'getNotificationChannels', [])),
  getNotificationSettings: () => new Promise((resolve, reject) => exec(resolve, reject, 'UtilityPlugin', 'getNotificationSettings', [])),
  getVersionInfo: () => new Promise((resolve, reject) => exec(resolve, reject, 'UtilityPlugin', 'getVersionInfo', [])),
  openNotificationSettings: () => exec(null, null, 'UtilityPlugin', 'openNotificationSettings', []),
  getSimInfo: () => new Promise((resolve, reject) => exec(resolve, reject, 'UtilityPlugin', 'getSimInfo', []))
};

src/UtilityPlugin.java
package com.davidbriglio.utility;

import android.content.Context;
import android.content.Intent;
import android.telephony.TelephonyManager;
import android.app.NotificationManager;
import androidx.core.app.NotificationManagerCompat;
import android.app.NotificationChannel;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;
import java.util.List;

public class UtilityPlugin extends CordovaPlugin {
  @Override
  public boolean execute(final String action, final JSONArray args, final CallbackContext callbackContext)
      throws JSONException {
    Context mContext = this.cordova.getActivity().getApplicationContext();

    if (action.equals("getNotificationChannel") && android.os.Build.VERSION.SDK_INT >= 26) {
      NotificationManager manager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
      String channelId = args.getString(0);
      if (channelId != null && !channelId.equals("") && manager != null) {
        NotificationChannel channel = manager.getNotificationChannel(channelId);
        JSONObject payload = new JSONObject();

        if (channel != null) {
          // Put all channel settings into JSON to return to cordova
          payload.put("canBypassDnd", channel.canBypassDnd());
          payload.put("canShowBadge", channel.canShowBadge());
          payload.put("description", channel.getDescription());
          payload.put("group", channel.getGroup());
          payload.put("id", channel.getId());
          payload.put("importance", channel.getImportance());
          payload.put("lightColour", channel.getLightColor());
          payload.put("lockscreenVisibility", channel.getLockscreenVisibility());
          payload.put("name", channel.getName());
          payload.put("sound", channel.getSound());
          payload.put("vibrationPattern", channel.getVibrationPattern());
          payload.put("showsLights", channel.shouldShowLights());
          payload.put("vibration", channel.shouldVibrate());
          payload.put("summary", channel.toString());

          // if (android.os.Build.VERSION.SDK_INT >= 29) {
          // payload.put("canBubble", channel.canBubble());
          // payload.put("userSetImportance", channel.hasUserSetImportance());
          // }

          // if (android.os.Build.VERSION.SDK_INT >= 30) {
          // payload.put("conversationId", channel.getConversationId());
          // payload.put("parentChannelId", channel.getParentChannelId());
          // payload.put("userSetSound", channel.hasUserSetSound());
          // payload.put("importantConversation", channel.isImportantConversation());
          // }

          callbackContext.success(payload);
        } else {
          callbackContext.error("Notification channel not found.");
        }

      } else {
        callbackContext.error("Invalid channel id provided.");
      }
    } else if (action.equals("getNotificationChannels") && android.os.Build.VERSION.SDK_INT >= 26) {
      NotificationManager manager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
      List<NotificationChannel> channels = manager.getNotificationChannels();
      JSONArray payload = new JSONArray();

      for (int i = 0; i < channels.size(); i++) {
        payload.put(channels.get(i).getId());
      }

      callbackContext.success(payload);
    } else if (action.equals("getNotificationSettings")) {
      NotificationManagerCompat manager = NotificationManagerCompat.from(mContext);
      JSONObject payload = new JSONObject();

      // Get channel info
      if (manager != null) {
        payload.put("enabled", manager.areNotificationsEnabled());
        payload.put("importance", manager.getImportance());
      }

      callbackContext.success(payload);
    } else if (action.equals("getVersionInfo")) {
      JSONObject payload = new JSONObject();

      payload.put("sdk", android.os.Build.VERSION.SDK_INT);
      payload.put("baseOs", android.os.Build.VERSION.BASE_OS);
      payload.put("codeName", android.os.Build.VERSION.CODENAME);
      payload.put("incremental", android.os.Build.VERSION.INCREMENTAL);
      payload.put("previewSdk", android.os.Build.VERSION.PREVIEW_SDK_INT);
      payload.put("release", android.os.Build.VERSION.RELEASE);
      payload.put("securityPatch", android.os.Build.VERSION.SECURITY_PATCH);

      callbackContext.success(payload);
    } else if (action.equals("openNotificationSettings")) {
      Intent intent = new Intent();
      intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
      intent.setAction("android.settings.APP_NOTIFICATION_SETTINGS");

      //for Android 5-7
      intent.putExtra("app_package", mContext.getPackageName());
      intent.putExtra("app_uid", mContext.getApplicationInfo().uid);

      // for Android 8 and above
      intent.putExtra("android.provider.extra.APP_PACKAGE", mContext.getPackageName());

      mContext.startActivity(intent);
      callbackContext.success();
    } else if (action.equals("getSimInfo")) {
      JSONObject payload = new JSONObject();

      TelephonyManager telemamanger = (TelephonyManager) mContext.getSystemService(Context.TELEPHONY_SERVICE);

      try {
        payload.put("phoneNumber", telemamanger.getLine1Number());
      } catch (Exception e) {
        // We don't have permission
      }

      try {
        payload.put("serial", telemamanger.getSimSerialNumber());
        payload.put("imei", telemamanger.getDeviceId());
      } catch (Exception e) {
        // We don't have permission
      }

      callbackContext.success(payload);
    } else {
      callbackContext.error("Invalid plugin action");
    }

    return true;
  }
}
